rest_framework/fields.py

```diff
@@ -1270,6 +1274,70 @@ def to_internal_value(self, data):
         return data
 
 
+class RecursiveField(Field):
+    """
+    A field that gets its representation from its parent.
+
+    This method could be used to serialize a tree structure, a linked list, or
+    even a directed acyclic graph. As with all recursive things, it is
+    important to keep the base case in mind. In the case of the tree serializer
+    example below, the base case is a node with an empty list of children. In
+    the case of the list serializer below, the base case is when `next==None`.
+    Above all, beware of cyclical references.
+
+    Examples:
+
+    class TreeSerializer(self):
+        children = ListField(child=RecursiveField())
+
+    class ListSerializer(self):
+        next = RecursiveField(allow_null=True)
+    """
+
+    def __init__(self, to='self', to_module=None, **kwargs):
+        self.to = to
+        self.to_module = to_module
+        field_kwargs = dict(
+            (key, kwargs[key])
+            for key in kwargs
+            if key in inspect.getargspec(Field.__init__)
+        )
```

Comment:
There are a few major hurdles I had to figure out to get this to work, I think it's work enumerating them.

1. In order to support 'self' recursive serialization, I need to walk the list of parent references to get the parent serializer -- the earliest point at which this information is available is when the field is fully bound. If we drop 'self' recursive serialization and only allow people to do explicit imports by name, then this restriction is removed and I can create the proxy at any time after the proxied class is created.
2. In order to get validation to work, I had to either duplicate or forward various attributes on the `RecursiveField` to the proxied field. 
   1. My first, broad-sweeping attempt was to use `__getattribute__` to forward everything to the proxy after binding. This would cause the RecursiveField to look and behave exactly like the field it was imitating. This proved too be too effective, to the point of being confusing. It would forward the `__class__` attribute and the `fields` attribute, which would cause the `deepcopy` operation to create an object which was not a `RecursiveField` anymore, and caused `__repr__` to enter an infinite loop (since it recursively searches the field hierarchy).
   2. My second attempt involved not forwarding internal magic methods (things starting and ending with double underscores), and not forwarding the `fields` attribute, but I didn't like this because I didn't like having to make piecemeal exceptions
   3. I settled on:
      1. only forwarding bound instance methods, and 
      2. calling `__init__` and `bind` with the exact same arguments for both the `RecursiveField` and the proxied field to make sure that attributes were defined in both places.
   4. I am OK with forwarding the instance methods individually by name (`to_representation`, `to_internal_value`, `run_validators` etc), and we could do the same with the attributes (`read_only`, `write_only`, `field_name`...), but it feels somewhat fragile, since you will need to modify the recursive serializer any time you add or remove an attribute or method from `Field`